File: qt_demo/GoGame/GoSGF.py
class GoSGF(object):
    def __init__(self, f):
        self.len = None
        self.fileContext = f.readline()
        self.gameInformation = self.fileContext[2:-1].split(';')
        self.RU = None
        self.RE = None
        self.KM = None
        self.nowStep = 0
        self.steps = []
        self.initGlobalInformation()
        self.initSteps()

    def srcFile(self):
        return self.fileContext

    def initGlobalInformation(self):
        globalInformation = self.gameInformation[0]
        RUPos = globalInformation.find('RU')
        REPos = globalInformation.find('RE')
        KMPos = globalInformation.find('KM')
        self.RU = globalInformation[RUPos + 3:RUPos + 4] if RUPos != -1 else None
        self.RE = globalInformation[REPos + 3:REPos + 4] if REPos != -1 else None
        self.KM = float(globalInformation[KMPos + 3:KMPos + 4]) if KMPos != -1 else None

    def initSteps(self):
        srcSteps = self.gameInformation[1:]
        letterToNumber = {}
        for i in range(19):
            letter = 'a'
            letterToNumber[chr(ord(letter) + i)] = i
        for step in srcSteps:
            self.steps.append(Step(step, letterToNumber))
        self.len = len(self.steps)

    def show(self):
        print('RU :\t{:}'.format(self.RU))
        print('KM :\t{:}'.format(self.KM))
        print('result :\t{:}'.format(self.RE))
        for step in self.steps:
            print('{:} {:} {:}'.format(step.player, step.x, step.y))

    def haveNextStep(self):
        return self.nowStep < self.len

    def getNextStep(self):
        self.nowStep += 1
        return self.steps[self.nowStep - 1]


class Step(object):
    def __init__(self, srcString, letterToNumber):
        self.player = srcString[0]
        x = srcString[3]
        y = srcString[2]
        if 'a' <= x <= 'z' and 'a' <= y <= 'z':
            self.x = letterToNumber[srcString[3]]
            self.y = letterToNumber[srcString[2]]
        else:
            self.x = None
            self.y = None

File: MartrixGoC++/src/GoSGF.cpp
//
// Created by 11409 on 2021/4/21.
//

#include "GoSGF.hpp"
#include <iostream>

GoSGF::GoSGF(std::string &srcFileContext)
{
    this->nowStep = 0;
    this->fileContext = srcFileContext;
    this->gameInformation = this->fileContext.substr(INFORMATION_START, this->fileContext.size() - INFORMATION_END);
    this->RU = this->gameInformation[this->gameInformation.find("RU") + INFORMATION_BIAS];
    char tmpPlayer = this->gameInformation[this->gameInformation.find("RE") + INFORMATION_BIAS];
    this->RE = tmpPlayer == 'B' ? BLACK_PLAYER : WHITE_PLAYER;
    this->HA = this->gameInformation[this->gameInformation.find("HA") + INFORMATION_BIAS] - '0';
    int startOfStep = int(this->gameInformation.find(';'));
    if (this->HA > 1)
    {
        int afterABPos = int(this->gameInformation.find("AB")) + 2;
        int numOfHandCap = 0;
        for (int i = afterABPos; i < int(this->gameInformation.size()); i++)
        {
            if (this->gameInformation[i] == ';')
            {
                startOfStep = i;
                break;
            }
            if (this->gameInformation[i] == '[')
            { numOfHandCap++; }
        }
        for (int i = 0; i < numOfHandCap; i++)
        {
            this->steps.push_back(new Step('B',
                                           this->gameInformation[afterABPos + i * HAND_CAP_BIAS + 2],
                                           this->gameInformation[afterABPos + i * HAND_CAP_BIAS + 1]));
        }
    }
    bool comment = false;
    int length = int(this->gameInformation.size());
    for (int i = startOfStep; i < length; i++)
    {
        char c = this->gameInformation[i];
        if (comment && c != ']')
        { continue; }
        switch (c)
        {
            case 'C':
                comment = true;
                break;
            case ']':
            {
                if (comment)
                { comment = false; }
                break;
            }
            case 'W':
            case 'B':
                this->steps.push_back(new Step(c, this->gameInformation[i + X_BIAS], this->gameInformation[i + Y_BIAS]));
                break;
            default:
                break;
        }
    }
}

void GoSGF::show()
{
    std::cout << "Rule :\t" << this->RU << std::endl;
    std::cout << "result :\t" << (this->RE == BLACK_PLAYER ? "B" : "W") << std::endl;
    std::cout << "hand cap :\t" << this->HA << std::endl;
    for (auto &step : this->steps)
    {
        std::cout << (BLACK_PLAYER == step->player ? "B " : "W ") << *step;
    }
}

bool GoSGF::haveNextStep() const
{
    if (this->nowStep == int(this->steps.size()))
    {
        return false;
    } else
    {
        return true;
    }
}

void GoSGF::getNextStep(Step*&nextStep)
{
    nextStep = this->steps[this->nowStep];
    this->nowStep++;
}

GoSGF::~GoSGF()
{
    for(auto &step : this->steps)
    {
        delete step;
    }
}